internal/cri/server/container_image_mount.go

```diff
@@ -0,0 +1,212 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package server
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	containerd "github.com/containerd/containerd/v2/client"
+	"github.com/containerd/containerd/v2/core/leases"
+	"github.com/containerd/containerd/v2/core/mount"
+	"github.com/containerd/containerd/v2/defaults"
+	"github.com/containerd/errdefs"
+	"github.com/containerd/log"
+	"github.com/containerd/platforms"
+	"github.com/opencontainers/image-spec/identity"
+	imagespec "github.com/opencontainers/image-spec/specs-go/v1"
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+const (
+	labelGCSnapRef = "containerd.io/gc.ref.snapshot."
+)
+
+func (c *criService) mutateMounts(
+	ctx context.Context,
+	extraMounts []*runtime.Mount,
+	snapshotter string,
+	sandboxID string,
+	platform imagespec.Platform,
+
+) error {
+	for _, m := range extraMounts {
+		err := c.mutateImageMount(ctx, m, snapshotter, sandboxID, platform)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *criService) mutateImageMount(
+	ctx context.Context,
+	extraMount *runtime.Mount,
+	snapshotter string,
+	sandboxID string,
+	platform imagespec.Platform,
+) (retErr error) {
+	imageSpec := extraMount.GetImage()
+	if imageSpec == nil {
+		return nil
+	}
+	if extraMount.GetHostPath() != "" {
+		return fmt.Errorf("hostpath must be empty while mount image: %+v", extraMount)
+	}
+	if !extraMount.GetReadonly() {
+		return fmt.Errorf("readonly must be true while mount image: %+v", extraMount)
+	}
+
+	ref := imageSpec.GetImage()
+	if ref == "" {
+		return fmt.Errorf("image not specified in: %+v", imageSpec)
+	}
+	image, err := c.LocalResolve(ref)
+	if err != nil {
+		return fmt.Errorf("failed to resolve image %q: %w", ref, err)
+	}
+	containerdImage, err := c.toContainerdImage(ctx, image)
+	if err != nil {
+		return fmt.Errorf("failed to get image from containerd %q: %w", image.ID, err)
+	}
+
+	// This is a digest of the manifest
+	imageID := containerdImage.Target().Digest.Hex()
```

Comment:
hex is deprecated. https://github.com/opencontainers/go-digest/blob/1e56c6daea3b29090ddd17e3e9d88196f0d8815b/digest.go#L150-L153

btw, why do we use the manifest digest Hex as imageID?

Reply:
The mount path can be shared between multiple containers of the same pod.